File: plugins/plugin-matrix/rust/src/providers/user_context.rs
//! User context provider for Matrix plugin.

use crate::types::{get_matrix_localpart, get_matrix_user_display_name, MatrixUserInfo};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// User context data.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserContextData {
    pub user_id: String,
    pub display_name: String,
    pub localpart: String,
    pub avatar_url: Option<String>,
}

/// User context provider result.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserContextProviderResult {
    pub data: UserContextData,
    pub values: HashMap<String, serde_json::Value>,
    pub text: String,
}

/// Get the current Matrix user context.
pub fn get_user_context(
    sender: &MatrixUserInfo,
    agent_name: &str,
) -> UserContextProviderResult {
    let display_name = get_matrix_user_display_name(sender);
    let localpart = get_matrix_localpart(&sender.user_id).to_string();

    let response_text = format!(
        "{} is talking to {} ({}) on Matrix.",
        agent_name, display_name, sender.user_id
    );

    let data = UserContextData {
        user_id: sender.user_id.clone(),
        display_name: display_name.clone(),
        localpart: localpart.clone(),
        avatar_url: sender.avatar_url.clone(),
    };

    let mut values = HashMap::new();
    values.insert("user_id".to_string(), serde_json::json!(&sender.user_id));
    values.insert("display_name".to_string(), serde_json::json!(&display_name));
    values.insert("localpart".to_string(), serde_json::json!(&localpart));

    UserContextProviderResult {
        data,
        values,
        text: response_text,
    }
}

File: plugins/plugin-matrix/typescript/src/providers/userContext.ts
/**
 * User context provider for Matrix plugin.
 */

import type { Provider, ProviderResult, IAgentRuntime, Memory, State } from "@elizaos/core";
import { MatrixService } from "../service.js";
import {
  MATRIX_SERVICE_NAME,
  getMatrixLocalpart,
  getMatrixUserDisplayName,
  type MatrixUserInfo,
} from "../types.js";

/**
 * Provider that gives the agent information about the Matrix user context.
 */
export const userContextProvider: Provider = {
  name: "matrixUserContext",
  description: "Provides information about the Matrix user in the current conversation",

  get: async (
    runtime: IAgentRuntime,
    message: Memory,
    state: State
  ): Promise<ProviderResult> => {
    // Only provide context for Matrix messages
    if (message.content.source !== "matrix") {
      return {
        data: {},
        values: {},
        text: "",
      };
    }

    const matrixService = (await runtime.getService(MATRIX_SERVICE_NAME)) as MatrixService | undefined;

    if (!matrixService || !matrixService.isConnected()) {
      return {
        data: {},
        values: {},
        text: "",
      };
    }

    const agentName = state?.agentName || "The agent";

    // Try to get sender info from message metadata
    const metadata = message.content.metadata as Record<string, unknown> | undefined;
    const senderInfo = metadata?.senderInfo as MatrixUserInfo | undefined;

    if (!senderInfo) {
      return {
        data: {},
        values: {},
        text: "",
      };
    }

    const displayName = getMatrixUserDisplayName(senderInfo);
    const localpart = getMatrixLocalpart(senderInfo.userId);

    const responseText = `${agentName} is talking to ${displayName} (${senderInfo.userId}) on Matrix.`;

    return {
      data: {
        userId: senderInfo.userId,
        displayName,
        localpart,
        avatarUrl: senderInfo.avatarUrl,
      },
      values: {
        userId: senderInfo.userId,
        displayName,
        localpart,
      },
      text: responseText,
    };
  },
};
